### PR TITLE
chore: expand coding standards

### PR DESCRIPTION
**Pull request by jelbourn**
All aboard!
@kara @mmalerba @andrewseguin @tinayuangao @crisbeto @EladBezalel @topherfangio @DevVersion @josephperrott 

The coding standards were last updated just short of one year ago. Going forward, I want to update this doc much more regularly with insights that come out of code reviews. 

Are there any conventions that we've been enforcing through code review that are not part of this doc?

### CODE REVIEW

File: CODING_STANDARDS.md [hunk 1]
```diff
@@ -68,9 +72,97 @@ prefer sticking to a _single_ API for accomplishing something.
...
+be part of the user-facing API. This typically applies to symbols used in template expressions, 
+`@ViewChildren` / `@ContentChildren` properties, host bindings, and `@Input` / `@Output` properties
+(when using an alias).
+
+#### JsDoc comments
+
+All public APIs must have user-facing comments. These are extracted and shown in the documation
+on [material.angular.io](https://material.angular.io).
+
+Private and internal APIs should have JsDoc when they are not obvious. Ultimately it is the purview
```
Comment (crisbeto) [inline CODING_STANDARDS.md]: We should probably mention the `@docs-private` directive for excluding APIs that shouldn't be user-facing.

File: CODING_STANDARDS.md [hunk 2]
```diff
@@ -68,9 +72,97 @@ prefer sticking to a _single_ API for accomplishing something.
...
+of the code reviwer as to what is "obvious", but the rule of thumb is that *most* classes, 
+properties, and methods should have a JsDoc description. 
+
+Properties should have a concise description of what the property means:
+```ts
+  /** The label position relative to the checkbox. Defaults to 'after' */
+  @Input() labelPosition: 'before' | 'after' = 'after';
```
Comment (crisbeto) [inline CODING_STANDARDS.md]: Perhaps, in general, we should be turning these into types? E.g. `type MdLabelPosition = 'before' | 'after';` in order to allow users to annotate their code?

File: CODING_STANDARDS.md [hunk 3]
```diff
@@ -68,9 +72,97 @@ prefer sticking to a _single_ API for accomplishing something.
 
 ### TypeScript
 
-#### Provide function descriptions
-For functions that are more complicated than a simple getter/setter, provide at least a brief 
-sentence explaining what the function does and/or _why_ it does something.
+#### Typing
```
Comment (crisbeto) [inline CODING_STANDARDS.md]: We should mention that, even though TypeScript infers it, we should include the method's return type for Dgeni's sake.

File: CODING_STANDARDS.md [hunk 4]
```diff
@@ -68,9 +72,97 @@ prefer sticking to a _single_ API for accomplishing something.
...
+  /** The label position relative to the checkbox. Defaults to 'after' */
+  @Input() labelPosition: 'before' | 'after' = 'after';
+``` 
+
+Methods blocks should describe what the function does and provide a description for each parameter
+and the return value:
+```ts
+  /**
+   * Opens a modal dialog containing the given component.
+   * @param component Type of the component to load into the dialog.
```
Comment (EladBezalel) [inline CODING_STANDARDS.md]: We don't want to mention the parameter type?

Comment (crisbeto) [inline CODING_STANDARDS.md]: The type is already in the TypeScript annotations. AFAIK Dgeni picks those up properly.

**Comment by DevVersion** [comment]
What about mentioning the `100-column` limit, because the [Style Guidelines](https://google.github.io/styleguide/jsguide.html#formatting-column-limit) mention a 80-column limit, while we have 100-columns in our `tslint.json`.

File: CODING_STANDARDS.md [hunk 5]
```diff
@@ -68,9 +72,97 @@ prefer sticking to a _single_ API for accomplishing something.
...
+
+#### Host bindings
+Prefer using the `host` object in the directive configuration instead of `@HostBinding` and 
+`@HostListener`. We do this because TypeScript preserves the type information of methods with 
+decorators, and when one of the arguments for the method is a native `Event` type, this preserved
+type information can lead to runtime errors in non-browser environments (e.g., server-side 
+pre-rendering). 
+
 
 ### CSS
```
Comment (mmalerba) [inline CODING_STANDARDS.md]: I agree with most of this, but the "don't use flex" part has always bothered me. Do we have an example of this weird baseline behavior that we're afraid of? It's hard to tell exactly what the issue is from reading the spec...

File: CODING_STANDARDS.md [hunk 6]
```diff
@@ -68,9 +72,97 @@ prefer sticking to a _single_ API for accomplishing something.
...
+class UniqueSelectionDispatcher { }
+```
+
+Avoid suffixing a class with "Service", as it communicates nothing about what the class does. Try to
+think of the class name as a person's job title.
+
+##### Methods
+The name of a method should capture the action that is performed *by* that method.
+
+### Angular
```
Comment (mmalerba) [inline CODING_STANDARDS.md]: Should we mention the forRoot replacement solution?

File: CODING_STANDARDS.md (hunk 2 above)
Comment (jelbourn) [inline CODING_STANDARDS.md]: Probably, but I don't think that is general enough for the coding guidelines.

File: CODING_STANDARDS.md (hunk 6 above)
Comment (jelbourn) [inline CODING_STANDARDS.md]: I'm not sure that's general enough to belong in the coding guidelines.

File: CODING_STANDARDS.md (hunk 5 above)
Comment (jelbourn) [inline CODING_STANDARDS.md]: [One from the archives!](https://github.com/angular/material/issues/3234)

TL;DR The text baseline `display: flex` and `inline-flex` is computed from the first child of the flex container, which is different from `display: block` and `inline-block`. If you put two of these next to each other, the browser will align the baselines and the elements will be misaligned. 

Overall I think the messaging of "be cautious" with the two specific cases is good.

**Comment by jelbourn** [comment]
Addressed comments

File: CODING_STANDARDS.md [hunk 7]
```diff
@@ -58,19 +59,121 @@ Keeping modules to a single responsibility makes the code easier to test, consum
...
 
-#### Provide function descriptions
-For functions that are more complicated than a simple getter/setter, provide at least a brief 
-sentence explaining what the function does and/or _why_ it does something.
+#### Typing
+Avoid `any` where possible. If you find yourself using `any`, consider whether a generic may be
+appropriate in your case.
+
+For methods and properties that are part of a component's public API, all types must be explicitly
+specified because our documentation tooling cannot current infer types in places where TypeScript
```
Comment (topherfangio) [inline CODING_STANDARDS.md]: "cannot current infer" -> "cannot currently infer"

File: CODING_STANDARDS.md [hunk 8]
```diff
@@ -58,19 +59,121 @@ Keeping modules to a single responsibility makes the code easier to test, consum
...
+* Use `private` when appropriate and possible, prefixing the name with an underscore.
+* Prefix *library-internal* properties and methods with an underscore without using the `private` 
+keyword. This is necessary for anything that must be public (to be used by Angular), but should not
+be part of the user-facing API. This typically applies to symbols used in template expressions, 
+`@ViewChildren` / `@ContentChildren` properties, host bindings, and `@Input` / `@Output` properties
+(when using an alias).
+
+Additionally, the `@docs-private` JsDoc annotation can be used to hide any symbol from the public
+API docs.
+
```
Comment (topherfangio) [inline CODING_STANDARDS.md]: What about the `protected` keyword? Do we prefer to make it public instead? A small note about best practices may be beneficial.
